monkey/infection_monkey/monkey.py

```diff
@@ -204,11 +240,32 @@ def collect_system_info_if_configured(self):
         except Exception as e:
             logger.exception(f"Exception encountered during system info collection: {str(e)}")
 
-    def shutdown_by_not_alive_config(self):
-        if not WormConfiguration.alive:
-            raise PlannedShutdownException("Marked 'not alive' from configuration.")
+    def _start_propagation(self):
+        if not InfectionMonkey._max_propagation_depth_reached():
+            logger.info("Starting the propagation phase.")
+            logger.debug("Running with depth: %d" % WormConfiguration.depth)
+            self._propagate()
+        else:
+            logger.info("Maximum propagation depth has been reached; monkey will not propagate.")
+            TraceTelem(MAX_DEPTH_REACHED_MESSAGE).send()
+
+        if self._keep_running and WormConfiguration.alive:
+            InfectionMonkey._run_ransomware()
+
+        # if host was exploited, before continue to closing the tunnel ensure the exploited
+        # host had its chance to connect to the tunnel
+        if len(self._exploited_machines) > 0:
+            time_to_sleep = WormConfiguration.keep_tunnel_open_time
+            logger.info(
+                "Sleeping %d seconds for exploited machines to connect to tunnel", time_to_sleep
+            )
+            time.sleep(time_to_sleep)
```

Comment:
Based on comments this should be a method named something like `wait_for_victim_connection_to_tunnel`

Reply:
We minimize changes to the legacy code. Master should handle this.